include/fmt/core.h

```diff
@@ -1613,6 +1629,194 @@ inline void print(const S& format_str, Args&&... args) {
       internal::make_args_checked<Args...>(format_str, args...));
 #endif
 }
+
+namespace internal {
+
+template <typename T, typename Char> struct is_string_view : std::false_type {};
+
+template <typename Char>
+struct is_string_view<basic_string_view<Char>, Char> : std::true_type {};
+
+#ifdef FMT_USE_STRING_VIEW
+template <typename Traits, typename Char>
+struct is_string_view<std::basic_string_view<Char, Traits>, Char>
+    : std::true_type {};
+#endif
+
+#ifdef FMT_USE_EXPERIMENTAL_STRING_VIEW
+template <typename Traits, typename Char>
+struct is_string_view<std::experimental::basic_string_view<Char, Traits>, Char>
+    : std::true_type {};
+#endif
+
+template <typename T> struct is_ref_wrapper : std::false_type {};
+
+template <typename T>
+struct is_ref_wrapper<std::reference_wrapper<T>> : std::true_type {};
+
+template <typename T, typename Context> struct need_dyn_copy {
+  using mapped_type = mapped_type_constant<T, Context>;
+  static_assert(mapped_type::value != internal::type::named_arg_type,
+                "Bug indicator. Named arguments must be processed separately");
+
+  using type = std::integral_constant<
+      bool, !(is_ref_wrapper<T>::value ||
+              is_string_view<T, typename Context::char_type>::value ||
+              (mapped_type::value != internal::type::cstring_type &&
+               mapped_type::value != internal::type::custom_type &&
+               mapped_type::value != internal::type::string_type))>;
+};
+
+template <typename T, typename Context>
+using need_dyn_copy_t = typename need_dyn_copy<T, Context>::type;
+
+class dyn_arg_storage {
+  // Workaround clang's -Wweak-vtables. For templates (unlike regular classes
+  // doesn't complain about inability to deduce translation unit to place vtable
+  // So dyn_arg_node_base is made a fake template
+
+  template <typename = void> struct storage_node_base {
+    using owning_ptr = std::unique_ptr<storage_node_base<>>;
+    virtual ~storage_node_base() = default;
+    owning_ptr next_;
+  };
+
+  using owning_ptr = storage_node_base<>::owning_ptr;
+
+  template <typename T> struct storage_node : storage_node_base<> {
+    T value_;
+    FMT_CONSTEXPR explicit storage_node(const T& arg, owning_ptr&& next)
+        : value_{arg} {
+      // Must be initialised after value_
```

Comment:
Why?

Reply:
Because copying of arg may throw. If by that time we transfer ownership... you know.